manager/scheduler/constraint.go

```diff
@@ -44,26 +44,36 @@ func (f *ConstraintFilter) Check(n *NodeInfo) bool {
 			if !constraint.Match(n.Description.Hostname) {
 				return false
 			}
+		case "node.role":
+			if !constraint.Match(n.Spec.Role.String()) {
+				return false
+			}
 		default:
 			// default is node label in form like 'node.labels.key==value'
 			// if it is not well formed, always fails it
 			if !strings.HasPrefix(constraint.Key, "node.labels.") {
 				return false
 			}
-			// if the node doesn't have any label,
-			// it's equivalent to match an empty value.
-			// that is, 'node.labels.key!=value' should pass and
-			// 'node.labels.key==value' should fail
-			if n.Spec.Annotations.Labels == nil {
-				if !constraint.Match("") {
-					return false
+
+			// node has labels from node.Spec.Annotations.Labels and node.Description.Engine.Labels
+			// if a label exists on both, node.Spec.Annotations.Labels has precedence
+			combinedLabels := make(map[string]string)
+			if n.Description != nil && n.Description.Engine != nil && n.Description.Engine.Labels != nil {
```

Comment:
What I want is arbitrary constraints over a domain that is controller by users (through potentially signed specs).

If I can make deployment constraints that I can ensure can't be faked by a worker, that meets the objective. So for example, if I can define that `AppA` only runs on `--constraint node.labels.key==pci-dss`, and we can give guarantees that a worker can never change its own `node.labels`, then I can make `AppA` have `pci-dss` as a constraint, and be assured that a manager will never deploy it in an "insecure" hosts (non-labeled).